syntax = "proto3";

package metatrader;

service SymbolsService {

    // Gets an array of symbol infos.
    rpc GetSymbolInfos (GetSymbolInfosReq) returns (SymbolInfosList);

    // Gets ticker information of a given symbol.
    rpc GetSymbolTicker (GetSymbolTickerReq) returns (SymbolTicker);
}

// https://www.mql5.com/en/docs/constants/environment_state/marketinfoconstants#enum_symbol_swap_mode
enum SwapMode {
    DISABLED = 0;
    POINTS = 1;
    CURRENCY_SYMBOL = 2;
    CURRENCY_MARGIN = 3;
    CURRENCY_DEPOSIT = 4;
    INTEREST_CURRENT = 5;
    INTEREST_OPEN = 6;
    REOPEN_CURRENT = 7;
    REOPEN_BID = 8;
}

enum DayOfWeek {
    SUNDAY = 0;
    MONDAY = 1;
    TUESDAY = 2;
    WEDNESDAY = 3;
    THURSDAY = 4;
    FRIDAY = 5;
    SATURDAY = 6;
}

message SymbolInfo {
    string symbol = 1;
    string baseCurrency = 2;
    string quotedCurrency = 3;
    float contractSize = 4;
    int32 digits = 5;
    float tickSize = 6;
    float volumeMin = 7;
    float volumeMax = 8;
    float volumeStep = 9;
    SwapMode swapMode = 10;
    float swapLong = 11;
    float swapShort = 12;
    float swap3days = 13;
}

message SymbolTicker {
    int64 time = 1;
    double ask = 2;
    double bid = 3;
    double last = 4;
    int64 volume = 5;
}

message GetSymbolInfosReq {
    optional string symbol = 1;
}

message SymbolInfosList {
    repeated SymbolInfo symbolInfos = 1;
}

message GetSymbolTickerReq {
    string symbol = 1;
}
